### ISSUE TITLE
More efficient getting best-path

### ISSUE THREAD

**Issue by danpovey** [k2-fsa/k2#528]
The below is part of an nsys profile trace of the decoding (I increased from max_frames=50000 to max_frames=200000).
It shows that getting the 1-best out of the best-path is taking about 10% of the total time, which it shouldn't.

I think it would be more efficient to first filter best_scores.aux_labels using RemoveValuesLeq(aux_labels, 0), and then get the best-paths by accessing the row-splits and values of the resulting `Ragged<int32_t>` as torch tensors.  This would avoid manipulating individual FSAs.  I think we should add a file python/csrc/torch/ragged_ops.cu where we wrap some of the things in ragged_ops.h, starting with RemoveValuesLeq().

<img width="1240" alt="Screen Shot 2020-12-19 at 5 14 37 PM" src="https://user-images.githubusercontent.com/3298747/102685767-e184e580-421d-11eb-8c65-eff7d1ab5970.png">

**Comment by danpovey** [comment]
.. incidentally, since it's wasteful to keep the epsilons in the aux_labels of the decoding graph, we could add a template RemoveValuesEqual() [like RemoveValuesLeq()] to ragged_ops_inl.h as well, and use it to remove the epsilons from the aux_labels when we create the decoding graph.   I'd suggest to have the user do this, as in `graph.aux_labels = k2.ragged_ops.remove_values_equal(aux_labels, 0)`, as there may be situations where the zeros are meaningful.

**Comment by csukuangfj** [comment]
working on this.


---

> It shows that getting the 1-best out of the best-path is taking about 10% of the total time, which it shouldn't

https://github.com/k2-fsa/snowfall/blob/1ebd8a1aa0f2da22953806e4511b8217e750b415/egs/librispeech/asr/simple_v1/decode.py#L71

```python
        for i in range(len(texts)):
            if isinstance(best_paths[i].aux_labels, torch.Tensor):
                aux_labels = best_paths[i].aux_labels
            else:
                # it's a ragged tensor
                aux_labels = best_paths[i].aux_labels.values()
```

It would be more efficient if we cache the result of `best_path[i]` so that `__getitem__` is invoked only once.

**Comment by danpovey** [comment]
Mm, agreed.  But I think there is much more speedup available by avoiding using that subsripting operator at all.

**Comment by csukuangfj** [comment]
Closed by #533
